File: android/app/src/main/java/technology/tiny/app/geo/Geo.kt
package technology.tiny.app.geo

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import androidx.core.content.ContextCompat
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import com.google.android.gms.tasks.CancellationTokenSource
import java.util.Locale
import kotlin.coroutines.resume
import kotlin.math.roundToInt
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull

/**
 * 📍 Device location → agent context (maps-location loop c6; web lib/geo.ts
 * parity, agi-diy context-injector grammar).
 *
 * The pure half (kmh / cardinal / contextBlock) matches the web's `###
 * Location` block byte-for-byte so every client teaches the tiny the same
 * shape. The FusedLocation half is snapshot-based like [Motion]: one
 * getCurrentLocation per ask (bounded, cached 30s), never a standing
 * watch — the chat toggle costs a GPS fix per message, not a duty cycle.
 */
object Geo {
    data class Fix(
        val lat: Double,
        val lng: Double,
        /** meters, null when the platform won't say */
        val accuracyM: Int? = null,
        /** meters above sea level */
        val altitudeM: Int? = null,
        /** m/s from the platform — converted at render time */
        val speedMs: Double? = null,
        /** degrees clockwise from true north, null when not moving */
        val headingDeg: Double? = null,
        val timestampMs: Long = 0,
    )

    private val CARDINALS = arrayOf("N", "NE", "E", "SE", "S", "SW", "W", "NW")

    /** m/s → km/h at 1dp; null for junk/negative (web speedKmh parity). */
    fun kmh(speedMs: Double?): Double? {
        if (speedMs == null || !speedMs.isFinite() || speedMs < 0) return null
        return (speedMs * 3.6 * 10).roundToInt() / 10.0
    }

    /** 0-360° → compass point; wraps, null for junk (web headingCardinal parity). */
    fun cardinal(deg: Double?): String? {
        if (deg == null || !deg.isFinite()) return null
        val norm = ((deg % 360.0) + 360.0) % 360.0
        return CARDINALS[((norm / 45.0).roundToInt()) % 8]
    }

    /**
     * The agent-facing `### Location` markdown block — same grammar as web
     * lib/geo.ts locationContext / agi-diy context-injector. Locale.US
     * pinned: a tr-TR device would otherwise print "37,7749" and split the
     * coordinate pair on its own comma.
     */
    fun contextBlock(fix: Fix?): String {
        if (fix == null || !fix.lat.isFinite() || !fix.lng.isFinite()) return ""
        val lines = mutableListOf("### Location")
        lines += "- **Coordinates**: ${"%.4f".format(Locale.US, fix.lat)}, ${"%.4f".format(Locale.US, fix.lng)}"
        fix.accuracyM?.let { lines += "- **Accuracy**: ±${it}m" }
        fix.altitudeM?.let { lines += "- **Altitude**: ${it}m" }
        kmh(fix.speedMs)?.takeIf { it > 0 }?.let {
            lines += "- **Speed**: ${"%.1f".format(Locale.US, it)} km/h"
        }
        cardinal(fix.headingDeg)?.let {
            lines += "- **Heading**: $it (${fix.headingDeg!!.roundToInt()}°)"
        }
        return lines.joinToString("\n")
    }

    fun hasPermission(context: Context): Boolean =
        ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED ||
            ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED

    fun fromLocation(loc: Location): Fix = Fix(
        lat = loc.latitude,
        lng = loc.longitude,
        accuracyM = if (loc.hasAccuracy()) loc.accuracy.roundToInt() else null,
        altitudeM = if (loc.hasAltitude()) loc.altitude.roundToInt() else null,
        speedMs = if (loc.hasSpeed()) loc.speed.toDouble() else null,
        headingDeg = if (loc.hasBearing()) loc.bearing.toDouble() else null,
        timestampMs = loc.time,
    )

    @Volatile private var cached: Fix? = null
    private const val CACHE_MS = 30_000L
    private const val FIX_TIMEOUT_MS = 5_000L

    /**
     * One bounded fix: 30s cache → fused getCurrentLocation (≤5s) → null.
     * Null on no-permission — callers degrade to "no location line", the
     * Motion/Bluetooth snapshot contract.
     */
    @SuppressLint("MissingPermission") // guarded by hasPermission
    suspend fun current(context: Context): Fix? {
        if (!hasPermission(context)) return null
        cached?.takeIf { System.currentTimeMillis() - it.timestampMs < CACHE_MS }?.let { return it }
        val fix = withTimeoutOrNull(FIX_TIMEOUT_MS) {
            suspendCancellableCoroutine<Fix?> { cont ->
                val cts = CancellationTokenSource()
                LocationServices.getFusedLocationProviderClient(context)
                    .getCurrentLocation(Priority.PRIORITY_BALANCED_POWER_ACCURACY, cts.token)
                    .addOnSuccessListener { loc -> if (cont.isActive) cont.resume(loc?.let(::fromLocation)) }
                    .addOnFailureListener { if (cont.isActive) cont.resume(null) }
                cont.invokeOnCancellation { cts.cancel() }
            }
        }
        if (fix != null) cached = fix
        return fix
    }

    /**
     * The per-send hook (ChatViewModel extraSystem): the location block when
     * the settings toggle is on AND permission is granted, else null and the
     * request is byte-identical to before — web mergeLocationMeta parity.
     */
    suspend fun contextIfEnabled(context: Context, enabled: Boolean): String? {
        if (!enabled) return null
        val block = contextBlock(current(context) ?: return null)
        return block.ifBlank { null }
    }
}

File: ios/Tiny/Sources/Geo.swift
/**
 * 📍 Geo — device location → agent context (maps-location loop c8).
 *
 * iOS port of web lib/geo.ts / Android geo/Geo.kt: the pure half (kmh /
 * cardinal / contextBlock) renders the exact same `### Location` markdown
 * block, so every client teaches the tiny one shape (String(format:) with
 * no locale is C-locale — no tr-TR comma-splitting the coordinate pair).
 *
 * The CoreLocation half is snapshot-based like Motion/Bluetooth: one
 * requestLocation per ask — 30s-cached, 5s-bounded, never a standing
 * watch. Permission is asked by the Settings toggle (requestPermission),
 * never implicitly; unauthorized just means "no location line".
 */
import CoreLocation
import Foundation

final class Geo: NSObject, CLLocationManagerDelegate, @unchecked Sendable {
    static let shared = Geo()

    struct Fix {
        let lat: Double
        let lng: Double
        /// meters, nil when the platform won't say
        let accuracyM: Int?
        /// meters above sea level
        let altitudeM: Int?
        /// m/s — nil when stationary/unknown (CLLocation sends -1)
        let speedMs: Double?
        /// degrees clockwise from true north; nil when not moving
        let headingDeg: Double?
        let timestampMs: Double
    }

    // ── pure half (byte-parity with web tests/geo.test.ts + GeoTest.kt) ──

    private static let cardinals = ["N", "NE", "E", "SE", "S", "SW", "W", "NW"]

    /// m/s → km/h at 1dp; nil for junk/negative.
    static func kmh(_ speedMs: Double?) -> Double? {
        guard let s = speedMs, s.isFinite, s >= 0 else { return nil }
        return (s * 3.6 * 10).rounded() / 10
    }

    /// 0-360° → compass point; wraps, nil for junk.
    static func cardinal(_ deg: Double?) -> String? {
        guard let d = deg, d.isFinite else { return nil }
        let norm = (d.truncatingRemainder(dividingBy: 360) + 360).truncatingRemainder(dividingBy: 360)
        return cardinals[Int((norm / 45).rounded()) % 8]
    }

    /// The agent-facing `### Location` block — same grammar on every client.
    static func contextBlock(_ fix: Fix?) -> String {
        guard let f = fix, f.lat.isFinite, f.lng.isFinite else { return "" }
        var lines = ["### Location"]
        lines.append("- **Coordinates**: \(String(format: "%.4f", f.lat)), \(String(format: "%.4f", f.lng))")
        if let a = f.accuracyM { lines.append("- **Accuracy**: ±\(a)m") }
        if let alt = f.altitudeM { lines.append("- **Altitude**: \(alt)m") }
        if let v = kmh(f.speedMs), v > 0 {
            lines.append("- **Speed**: \(String(format: "%.1f", v)) km/h")
        }
        if let c = cardinal(f.headingDeg), let deg = f.headingDeg {
            lines.append("- **Heading**: \(c) (\(Int(deg.rounded()))°)")
        }
        return lines.joined(separator: "\n")
    }

    /// CLLocation → Fix (CoreLocation's "invalid" sentinels → nil fields).
    static func fix(from loc: CLLocation) -> Fix {
        Fix(
            lat: loc.coordinate.latitude,
            lng: loc.coordinate.longitude,
            accuracyM: loc.horizontalAccuracy >= 0 ? Int(loc.horizontalAccuracy.rounded()) : nil,
            altitudeM: loc.verticalAccuracy > 0 ? Int(loc.altitude.rounded()) : nil,
            speedMs: loc.speed >= 0 ? loc.speed : nil,
            headingDeg: loc.course >= 0 ? loc.course : nil,
            timestampMs: loc.timestamp.timeIntervalSince1970 * 1000
        )
    }

    // ── CoreLocation half ──

    private let manager = CLLocationManager()
    private let lock = NSLock()
    private var pending: [UUID: CheckedContinuation<Fix?, Never>] = [:]
    private var cached: Fix?

    private static let cacheMs: Double = 30_000
    private static let fixTimeoutS: Double = 5

    override private init() {
        super.init()
        manager.delegate = self
        // Hundred-meter class: the context block is 4-decimal (~11m) anyway,
        // and coarse fixes answer in well under the 5s bound.
        manager.desiredAccuracy = kCLLocationAccuracyHundredMeters
    }

    var authorized: Bool {
        switch manager.authorizationStatus {
        case .authorizedWhenInUse, .authorizedAlways: return true
        default: return false
        }
    }

    /// The user said no (or a profile did) — as opposed to .notDetermined,
    /// where the ask dialog is still the next step. UIs use this to swap
    /// "waiting for position…" for the truth.
    var denied: Bool {
        switch manager.authorizationStatus {
        case .denied, .restricted: return true
        default: return false
        }
    }

    /// Settings-toggle hook — the ONLY place the system prompt fires.
    func requestPermission() {
        if manager.authorizationStatus == .notDetermined {
            manager.requestWhenInUseAuthorization()
        }
    }

    /// Sync (lockable — Swift 6 bars NSLock in async bodies): fresh cache or nil.
    private func freshCached() -> Fix? {
        lock.lock()
        defer { lock.unlock() }
        if let c = cached, Date().timeIntervalSince1970 * 1000 - c.timestampMs < Self.cacheMs {
            return c
        }
        return nil
    }

    /// One bounded fix: 30s cache → requestLocation (≤5s) → nil.
    func current() async -> Fix? {
        guard authorized else { return nil }
        if let c = freshCached() { return c }

        let id = UUID()
        return await withCheckedContinuation { (cont: CheckedContinuation<Fix?, Never>) in
            lock.lock()
            pending[id] = cont
            let first = pending.count == 1
            lock.unlock()
            if first {
                DispatchQueue.main.async { self.manager.requestLocation() }
            }
            // Timeout: whoever still owns the id resumes it — the id-keyed
            // map is the double-resume guard (delegate may have raced us).
            DispatchQueue.global().asyncAfter(deadline: .now() + Self.fixTimeoutS) {
                self.lock.lock()
                let c = self.pending.removeValue(forKey: id)
                self.lock.unlock()
                c?.resume(returning: nil)
            }
        }
    }

    /// The per-send hook (ChatController extraSystem): the location block
    /// when the Settings toggle is on AND permission granted, else nil and
    /// the request is byte-identical to before.
    func contextIfEnabled() async -> String? {
        guard Config.locationContext, authorized else { return nil }
        guard let f = await current() else { return nil }
        let block = Self.contextBlock(f)
        return block.isEmpty ? nil : block
    }

    private func flush(_ fix: Fix?) {
        lock.lock()
        if let f = fix { cached = f }
        let conts = pending
        pending = [:]
        lock.unlock()
        conts.values.forEach { $0.resume(returning: fix) }
    }

    // MARK: CLLocationManagerDelegate (nonisolated — flush() is lock-guarded)

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        flush(locations.last.map(Self.fix(from:)))
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        flush(nil)
    }
}
